eax/eax.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2019 ProtonTech AG
+
+// Package eax provides an implementation of the EAX
+// (encrypt-authenticate-translate) mode of operation, as described in
+// Bellare, Rogaway, and Wagner "THE EAX MODE OF OPERATION: A TWO-PASS
+// AUTHENTICATED-ENCRYPTION SCHEME OPTIMIZED FOR SIMPLICITY AND EFFICIENCY."
+// In FSE'04, volume 3017 of LNCS, 2004
+
+package eax
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"errors"
+	"golang.org/x/crypto/internal/byteutil"
+)
+
+const (
+	defaultTagSize   = 16
+	defaultNonceSize = 16
+)
+
+type eax struct {
+	block     cipher.Block // Only AES-{128, 192, 256} supported
+	tagSize   int          // At least 12 bytes recommended
+	nonceSize int
+}
+
+func (e *eax) NonceSize() int {
+	return e.nonceSize
+}
+
+func (e *eax) Overhead() int {
+	return e.tagSize
+}
+
+// NewEAX returns an EAX instance with AES-{keyLength} and default tag and
+// nonce lengths.
+func NewEAX(key []byte) (cipher.AEAD, error) {
+	return NewEAXWithNonceAndTagSize(key, defaultNonceSize, defaultTagSize)
+}
+
+// NewEAXWithNonceAndTagSize returns an EAX instance with AES-{keyLength} and
+// given nonce and tag lengths in bytes. Panics on zero nonceSize and
+// exceedingly long tags.
+//
+// It is recommended to use at least 12 bytes as tag length (see, for instance,
+// NIST SP 800-38D).
+//
+// Only to be used for compatibility with existing cryptosystems with
+// non-standard parameters. For all other cases, prefer NewEAX.
+func NewEAXWithNonceAndTagSize(
+	key []byte, nonceSize, tagSize int) (cipher.AEAD, error) {
+	if nonceSize < 1 {
+		return nil, eaxError("Cannot initialize EAX with nonceSize = 0")
+	}
+	aesCipher, err := aes.NewCipher(key)
+	if err != nil {
+		panic(err)
+	}
+	if tagSize > aesCipher.BlockSize() {
+		return nil, eaxError("Custom tag length exceeds blocksize")
+	}
+	return &eax{
+		block:     aesCipher,
+		tagSize:   tagSize,
+		nonceSize: nonceSize,
+	}, nil
+}
+
+// Seal (AEAD interface) returns a byte array containing the
+// concatenation of the ciphertext and the validation tag.
+func (e *eax) Seal(dst, nonce, plaintext, adata []byte) []byte {
+	if len(nonce) > e.nonceSize {
+		panic("crypto/eax: Incorrect nonce length given to EAX")
+	}
+	omacNonce := e.omacT(0, nonce)
+	omacAdata := e.omacT(1, adata)
+
+	// Encrypt message using CTR mode and omacNonce as IV
+	ctr := cipher.NewCTR(e.block, omacNonce)
+	ciphertext := make([]byte, len(plaintext))
+	ctr.XORKeyStream(ciphertext, plaintext)
+
+	omacCiphertext := e.omacT(2, ciphertext)
+
+	tag := make([]byte, e.tagSize)
+	for i := 0; i < e.tagSize; i++ {
+		tag[i] = omacCiphertext[i] ^ omacNonce[i] ^ omacAdata[i]
+	}
+
+	return append(ciphertext, tag...)
+}
+
+// Open (the AEAD interface) returns a byte array containing the plaintext and
+// the eventual authentication error.
+func (e* eax) Open(dst, nonce, ciphertext, adata []byte) ([]byte, error) {
+	if len(ciphertext) < e.tagSize {
+		return nil, eaxError("EAX: Ciphertext shorter than tag length")
+	}
```

Comment:
`eaxError` is nice from a DRY (don't repeat yourself) perspective, but it's nice to be able to `ctrl+f` the entire error message and find it, so I'd just spell it out here. (Also you'd end up with "crypto/eax: EAX: Ciphertext shorter than tag length" here which is not very DRY in itself 🙃)

Reply:
Yep, missed this one : - )
Do you mean I should not use eaxError at all because of the lazy `ctrl+f` or just fix this example?
I also did this in case the structure changes again, in order not to change every `crypto/eax` by something else each time.

Reply:
(and, by the way, [it does not seem that I can access the package name/location easily in Go](https://stackoverflow.com/questions/25262754/how-to-get-name-of-current-package-in-go))

Reply:
I meant the former, yeah. Go philosophy seems to be to be very explicit about these kinds of things, rather than introduce cleverness. I think renaming them when moving again is not so bad - if anything, it would be nice to move them into https://github.com/golang/go/tree/master/src/crypto/cipher, which is another repo anyway.

(If we could do that, we'd also have access to https://github.com/golang/go/blob/5498fa90e9921193be11c6e6a9e017a53c0e1588/src/crypto/cipher/xor_generic.go#L14, which'd be nice.)

Reply:
I am not very convinced by the `ctrl+f` argument. Personally, if I cant find the error message with grep, I'd almost instantaneously look for a substring. If that fails, I'd go to the `crypto/eax` package as shown in the error and look there. In my opinion, DRY outweighs here. 
The point that I was making is that _supposing we're moving those packages in the future_, which I agree is not bad, would involve changing every error message, something that can be easily overlooked, so that contributes to the point of not hardcoding error messages.